### PR TITLE
HPCC-26132 Fix problem with COUNTER defined as an attribute

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Test example is LNtesting/notactive2.eclxml

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -14262,6 +14262,7 @@ void exportData(IPropertyTree *data, IHqlExpression *table, bool flatten)
 
 static bool exprContainsCounter(RecursionChecker & checker, IHqlExpression * expr, IHqlExpression * counter)
 {
+    expr = expr->queryBody();
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp:14265]: This is the actual fix - if the counter was named it wasn't being matched

File: ecl/hql/hqlexpr.cpp [hunk 2]
```diff
@@ -16116,7 +16119,8 @@ IHqlExpression * createUniqueId(IAtom * name)
 static UniqueSequenceCounter counterSequence;
 IHqlExpression * createCounter()
 {
-    return createSequence(no_counter, makeIntType(8, false), NULL, counterSequence.next());
+    unique_id_t seq = counterSequence.next();
+    return createSequence(no_counter, makeIntType(8, false), NULL, seq);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp:16123]: changed to make it easier to add a debugging breakpoint

File: ecl/hql/hqlscope.cpp [hunk 3]
```diff
@@ -193,6 +193,7 @@ extern HQL_API void checkIndependentOfScope(IHqlExpression * expr)
         exprs.append(OLINK(scopeUsed.item(i)));
 
     EclIR::dbglogIR(exprs);
+    EclIR::dump_ir(exprs);
```
Comment (ghalliday) [inline ecl/hql/hqlscope.cpp:196]: this and next improves debug tracing

File: ecl/hql/hqlgram.y [hunk 4]
```diff
@@ -8901,8 +8901,7 @@ simpleDataSet
                         {
                             OwnedHqlExpr transform = parser->createRowAssignTransform($3, $6, $10);
                             $6.release();
-                            $7.release();
-                            $$.setExpr(createDataset(no_hqlproject, $3.getExpr(), createComma(transform.getClear(), $8.getExpr(), $10.getExpr())));
+                            $$.setExpr(createDataset(no_hqlproject, { $3.getExpr(), transform.getClear(), $8.getExpr(), $10.getExpr(), $7.getExpr() }));
```
Comment (ghalliday) [inline ecl/hql/hqlgram.y:8904]: unlikely to ever happen - but no reason not to support COUNTER from a record.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**